azukiapp/azk
# Several improvements related to virtual machine and the agent

This pull request brings major changes to how the `azk` agent handles the virtual machine, resulting in many improvements mainly (but not exclusively) to `azk` users using mac.

### Closed issues in this PR:

 - Closes #116, adding swap support through this change in the image of the virtual machine: https://github.com/azukiapp/debian2docker/commit/6e1c93392629dc412d5c6b26a1363623ac863f8d
 - Closes #305, this problem had to do with a failure in VirtualBox DHCP Server. If it was unable to provide an ip, that caused the start of other services (including docker) to fail. This was solved by removing the use of the DHCP Server system and substituting the fixed IP configuration system through  `guestproperty`, as can be seen in:: https://github.com/azukiapp/debian2docker/commit/885303981fd9a382a488b84f40ec099d3751e20f 
 - Closes #252, solving a critical bug that could remove the data disk and lose all images already downloaded, and also saved data in persistent mounts. A simple change which uses a symbolic link in place of the actual disc in VirtualBox prevents accidental removal of the disc, furthermore improvements have been made in the error treatment while removing the VirtualBox virtual machine.
 - Closes #277, the `azk agent stop` command now awaits for the agent exit to complete before giving a return. This prevents the situation where the user could try to start the agent right after finishing it and `azk` outputting that the agent would still be running.
 - Closes #297, with the improvements made in the treatment of the virtual machine removal and ip configuration, initial tests no longer accuse errors in the use of the virtual machine on Linux. But we can't be sure that the bug is fixed, and we're looking into it.

### Bonus:

Besides the problem with the DHCP Server, there was always a problem that could cause ip conflicts between the virtual machine and any other network interface (both real interfaces as other VirtualBox interfaces). In this PR, it was added an ip conflict treatment in the agent startup process, and also an ip suggestion system besides the default `192.168.50.4`, which should ease the configuration in the event of a conflict.

### How to test:

This update uses a new iso to run the virtual machine, so before running this change:

 - Stop the agent with `azk agent stop`;
 - Go to the folder `lib/vm` and rename the file `azk.iso` to `azk_stable.iso` (so you can change it back in case you need to test some other branch that does use this change);
 - Download the new iso with `azk nvm grunt vm-download`;

Note: Be careful when dealing with files inside `lib/vm`, especially when packaging `azk` for Mac. This folder should only contain the files that will really be packed;

### Thanks

Special thanks to @dynaum and @jturolla by patience and several reports on these issues, were of great value to solving the problem.
